File: pdoo/ruby/gameuniverse.rb
#encoding:utf-8

require "./lib/GameUniverseToUI.rb"
require "./lib/GameStateController.rb"
require_relative "./dice.rb"
require_relative "./spacestation.rb"
require_relative "./enemystarship"
require_relative "./gamecharacter.rb"
require_relative "./combatresult.rb"
require_relative "powerefficientspacestation"
require_relative "betapowerefficientspacestation"
require_relative "spacecity"
require_relative "transformation"

module Deepspace
  # Clase que desarrolla la partida
  class GameUniverse
    # Puntos necesarios para ganar
    @@WIN = 10
    
    def initialize
      @gameState = GameStateController.new
      @turns = 0
      @currentStationIndex = 0
      @dice = Dice.new
      @currentStation = nil
      @currentEnemy = nil
      @spaceStations = Array.new
      @haveSpaceCity = false
    end

    # Consultor del estado
    def state
      @gameState.state
    end

    # Se realiza un combate entre una estación espacial y un enemigo
    def combatGo(station,enemy)
      ch = @dice.firstShot()
      # Empieza el enemigo
      if ch==GameCharacter::ENEMYSTARSHIP
        fire = enemy.fire
        result = station.recieveShot(fire)
        
        if result==ShotResult::RESIST
          fire = station.fire
          result = enemy.recieveShot(fire)
          enemyWins = (result==ShotResult::RESIST)
        else
          enemyWins = true
        end
      # Empieza la nave
      else
        fire = station.fire
        result = enemy.recieveShot(fire)
        enemyWins = (result==ShotResult::RESIST)
      end
      
      # Gana el enemigo
      if enemyWins
        s = station.getSpeed
        moves = @dice.spaceStationMoves(s)
        if !moves
          damage = enemy.damage
          station.setPendingDamage(damage)
          return CombatResult::ENEMYWINS
        else
          station.move
          return CombatResult::SPACESTATIONESCAPES
        end
      # No gana el enemigo
      else
        aLoot = enemy.loot
        transform = station.setLoot(aLoot)
        if transform==Transformation::GETEFFICIENT
          makeStationEfficient
          return CombatResult::STATIONWINSANDCONVERTS
        elsif transform==Transformation::SPACECITY
          createSpaceCity
          return CombatResult::STATIONWINSANDCONVERTS
        else
          return CombatResult::STATIONWINS
        end
      end
    end

    # Realiza el combate de un turno
    def combat
      state = @gameState.state
      if state==GameState::BEFORECOMBAT or state==GameState::INIT
        combatResult = combatGo(@currentStation,@currentEnemy)
        @gameState.next(@turns,@spaceStations.length)
      else
        combatResult = CombatResult::NOCOMBAT
      end
      return combatResult
    end
    
    # La nave del turno descarta el hangar
    def discardHangar
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.discardHangar
      end
    end

    # La nave del turno descarta un potenciador de escudo
    def discardShieldBooster(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.discardShieldBooster(i)
      end
    end

    # La nave del turno descarta un potenciador de escudo de su hangar
    def discardShieldBoosterInHangar(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.discardShieldBoosterInHangar(i)
      end
    end

    # La nave del turno descarta un arma
    def discardWeapon(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.discardWeapon(i)
      end
    end

    # La nave del turno descarta un arma de su hangar
    def discardWeaponInHangar(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.discardWeaponInHangar(i)
      end
    end

    def getUIversion
      GameUniverseToUI.new(@currentStation,@currentEnemy)
    end

    # Comprueba si la nave del turno ha ganado la partida
    def haveAWinner
      if @currentStation.nMedals>=@@WIN
        return true
      end
      return false
    end

    # Comienza la partida
    def init(names)
      state = @gameState.state
      if state==GameState::CANNOTPLAY
        dealer = CardDealer.instance
        names.each do |name|
          supplies = dealer.nextSuppliesPackage
          station = SpaceStation.new(name,supplies)
          @spaceStations << station
          
          nh = @dice.initWithNHangars
          nw = @dice.initWithNWeapons
          ns = @dice.initWithNShields
          lo = Loot.new(0,nw,ns,nh,0)
          station.setLoot(lo)
        end

        @currentStationIndex = @dice.whoStarts(@spaceStations.length)
        @currentStation = @spaceStations[@currentStationIndex]
        @currentEnemy = dealer.nextEnemy
        @gameState.next(@turns,@spaceStations.length)
      end
    end

    # Monta un escudo con índice 'i' en la nave
    def mountShieldBooster(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.mountShieldBooster(i)
      end
    end

    # Monta un arma con índice 'i' en la nave
    def mountWeapon(i)
      if @gameState.state==GameState::INIT or @gameState.state==GameState::AFTERCOMBAT
        @currentStation.mountWeapon(i)
      end
    end

    # Pasa el turno si se puede
    def nextTurn
      state = @gameState.state
      if state==GameState::AFTERCOMBAT
        stationState = @currentStation.validState
        if stationState
          @currentStationIndex = (@currentStationIndex+1) % @spaceStations.length
          @turns += 1
          @currentStation = @spaceStations[@currentStationIndex]
          @currentStation.cleanUpMountedItems
          dealer = CardDealer.instance
          @currentEnemy = dealer.nextEnemy
          @gameState.next(@turns,@spaceStations.length)
          return true
        end
        return false
      end
      return false
    end

    def to_s
      getUIversion.to_s
    end

    def makeStationEfficient
      if @dice.extraEfficiency
        @currentStation = BetaPowerEfficientSpaceStation.new(@currentStation)
	@spaceStations[@currentSpaceStationIndex] = @currentStation
      else
        @currentStation = PowerEfficientSpaceStation.new(@currentStation)
	@spaceStations[@currentSpaceStationIndex] = @currentStation
      end
    end

    def createSpaceCity
      if @haveSpaceCity==false
        collaborators = @SpaceStation.select {|station| station!=@currentStation}
        @currentStation = SpaceCity.new(@currentStation,collaborators)
        @haveSpaceCity = true
	@spaceStations[@currentSpaceStationIndex] = @currentStation
      end
    end
  end
end

File: pdoo/java/Deepspace/GameUniverse.java
/**
 * @author Manuel Gachs Ballegeer
 */
package Deepspace;

import java.util.ArrayList;

/**
 * @brief Clase que controla la partida
 */
public class GameUniverse {
    private final static int WIN = 10;
    
    private ArrayList<SpaceStation> spaceStations = new ArrayList<>();
    private SpaceStation currentStation;
    private EnemyStarShip currentEnemy;
    private GameStateController gameState;
    private Dice dice;
    private int currentStationIndex;
    private int turns;
    private boolean hasSpaceCity;
    
    public GameUniverse() {
        gameState = new GameStateController();
        dice = new Dice();
        turns = 0;
        hasSpaceCity = false;
    }
    
    CombatResult combat(SpaceStation station,EnemyStarShip enemy) {
        GameCharacter ch = dice.firstShot();
        boolean enemy_wins;
        float fire;
        ShotResult result; 
        
        if (ch==GameCharacter.ENEMYSTARSHIP) {
            fire = enemy.fire();
            result = station.recieveShot(fire);
            
            if (result==ShotResult.RESIST) {
                fire = station.fire();
                result = enemy.recieveShot(fire);
                enemy_wins = (result==ShotResult.RESIST);
            } else {
                enemy_wins = true;
            }
        } else {
            fire = station.fire();
            result = enemy.recieveShot(fire);
            enemy_wins = (result==ShotResult.RESIST);
        }
               
        if (enemy_wins) {
            float s = station.getSpeed();
            boolean moves = dice.spaceStationMoves(s);
            if (!moves) {
                Damage damage = enemy.getDamage();
                station.setPendingDamage(damage);
                return CombatResult.ENEMYWINS;
            } else {
                station.move();
                return CombatResult.STATIONESCAPES;
            }
        } else {
            Loot aLoot = enemy.getLoot();
            Transformation transformation = station.setLoot(aLoot);
            if (transformation==Transformation.GETEFFICIENT) {
                makeStationEfficient();
                return CombatResult.STATIONWINSANDCONVERTS;
            } else if (transformation==Transformation.SPACECITY) {
                createSpaceCity();
                return CombatResult.STATIONWINSANDCONVERTS;
            } else
            return CombatResult.STATIONWINS;
        }
    }
    
    public CombatResult combat() {
        GameState state = gameState.getState();
        CombatResult combat_result;
        
        if (state==GameState.BEFORECOMBAT || state==GameState.INIT) {
            combat_result = combat(currentStation,currentEnemy);
            gameState.next(turns,spaceStations.size());
        } else
            combat_result = CombatResult.NOCOMBAT;
        return combat_result;
    }
    
    public void discardHangar() {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.discardHangar();
    }
    
    public void discardShieldBooster(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.discardShieldBooster(i);
    }
    
    public void discardShieldBoosterInHangar(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.discardShieldBoosterInHangar(i);
    }
    
    public void discardWeapon(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.discardWeapon(i);
    }
    
    public void discardWeaponInHangar(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.discardWeaponInHangar(i);
    }
    
    public GameState getState() { return gameState.getState(); }
    
    public GameUniverseToUI getUIversion() { return new GameUniverseToUI(currentStation,currentEnemy); }
    
    public boolean haveAWinner() { return currentStation.getNMedals()==WIN; }
    
    public void init(ArrayList<String> names) {
        GameState state = gameState.getState();
        
        if (state==GameState.CANNOTPLAY) {
            CardDealer dealer = CardDealer.getInstance();
            
            for (String name: names) {
                SuppliesPackage supplies = dealer.nextSuppliesPackage();
                SpaceStation station = new SpaceStation(name,supplies);
                spaceStations.add(station);
                
                int nh = dice.initWithNHangars();
                int nw = dice.initWithNWeapons();
                int ns = dice.initWithNShields();
                Loot lo = new Loot(0,nw,ns,nh,0);
                station.setLoot(lo);
            }
            
            currentStationIndex = dice.whoStarts(spaceStations.size());
            currentStation = spaceStations.get(currentStationIndex);
            currentEnemy = dealer.nextEnemy();
            gameState.next(turns, spaceStations.size());
        }
    }
    
    public void mountShieldBooster(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.mountShieldBooster(i);
    }
    
    public void mountWeapon(int i) {
        if (gameState.getState()==GameState.INIT || gameState.getState()==GameState.AFTERCOMBAT)
            currentStation.mountWeapon(i);
    }
    
    public boolean nextTurn() {
        GameState state = gameState.getState();
        
        if (state==GameState.AFTERCOMBAT) {
            boolean stationState = currentStation.validState();
            if (stationState) {
                currentStationIndex = (currentStationIndex+1) % spaceStations.size();
                turns += 1;
                currentStation = spaceStations.get(currentStationIndex);
                currentStation.cleanUpMountedUnits();
                CardDealer dealer = CardDealer.getInstance();
                currentEnemy = dealer.nextEnemy();
                gameState.next(turns, spaceStations.size());
                return true;
            }
            return false;
        }
        return false;
    }
    
    @Override
    public String toString() {
        String aux = "Estación espacial actual:\n"+currentStation.toString();
        aux += "\nEnemigo actual:\n"+currentEnemy.toString();
        return aux;
    }
    
    public void createSpaceCity() {
        if (!hasSpaceCity) {
            ArrayList<SpaceStation> collaborators = new ArrayList<>();
            for (SpaceStation s: spaceStations)
                if (s!=currentStation)
                    collaborators.add(s);
            currentStation = new SpaceCity(currentStation,collaborators);
            hasSpaceCity = true;
        }
    }
    
    public void makeStationEfficient() {
        if (dice.extraEfficiency())
            currentStation = new BetaPowerEfficientSpaceStation(currentStation);
        else
            currentStation = new PowerEfficientSpaceStation(currentStation);
    }
}
